File: sdk/python/examples/61_github_coding_agent_chained.py
#!/usr/bin/env python3
# Copyright (c) 2025 Agentspan
# Licensed under the MIT License. See LICENSE file in the project root for details.

"""GitHub Coding Agent — issue to PR pipeline.

Deploys and serves a three-stage pipeline:
  1. Fetch open issue, create branch (CLI tools: gh, git)
  2. Code fix + QA review (SWARM: coder <-> qa_tester)
  3. Create pull request (CLI tool: gh)

Run:
    python github_coding_agent.py          # Deploy + serve
    agentspan run github_pipeline "..."    # Trigger (from another terminal)

Requirements:
    - Agentspan server running
    - GITHUB_TOKEN stored: agentspan credentials set --name GITHUB_TOKEN
    - gh CLI installed
"""

from agentspan.agents import Agent, AgentRuntime, Strategy
from agentspan.agents.gate import TextGate
from agentspan.agents.handoff import OnTextMention

REPO = "agentspan/codingexamples"
MODEL = "anthropic/claude-sonnet-4-6"

# ── Stage 1: Fetch issues ─────────────────────────────────────────

git_fetch_issues = Agent(
    name="git_fetch_issues",
    model=MODEL,
    instructions=f"""\
You are a GitHub issue fetcher.

1. List the 5 most recent open issues on {REPO}.
2. If there are NO open issues, output exactly: NO_OPEN_ISSUES
3. Otherwise pick the most suitable issue, then:
   - Create a temp dir, clone {REPO}, create branch fix/issue-<NUMBER>
   - Push the branch to origin
   - Output ONLY: REPO / BRANCH / ISSUE / SUMMARY lines
""",
    cli_commands=True,
    cli_allowed_commands=["gh", "git", "mktemp", "rm"],
    credentials=["GITHUB_TOKEN", "GH_TOKEN"],
    max_turns=20,
    gate=TextGate("NO_OPEN_ISSUES"),
)

# ── Stage 2: Coding + QA (SWARM) ──────────────────────────────────

coder = Agent(
    name="coder",
    model=MODEL,
    max_tokens=60000,
    credentials=["GITHUB_TOKEN", "GH_TOKEN"],
    instructions="""\
You are a senior developer. Clone the repo, check out the branch,
implement the fix, commit, push. Then say HANDOFF_TO_QA with REPO/BRANCH/CHANGES.
""",
    local_code_execution=True,
)

qa_tester = Agent(
    name="qa_tester",
    model=MODEL,
    credentials=["GITHUB_TOKEN", "GH_TOKEN"],
    instructions="""\
You are a QA engineer. Clone the repo, review changes, run tests.
If bugs found: say HANDOFF_TO_CODER with what to fix.
If good: say QA_APPROVED with REPO/BRANCH/SUMMARY.
""",
    local_code_execution=True,
    max_tokens=60000,
    max_turns=5,
)

coding_qa = Agent(
    name="coding_qa",
    model=MODEL,
    instructions=(
        "Delegate to coder, then qa_tester. Loop until QA approves. "
        "Output REPO/BRANCH/SUMMARY when done."
    ),
    agents=[coder, qa_tester],
    strategy=Strategy.SWARM,
    handoffs=[
        OnTextMention(text="HANDOFF_TO_QA", target="qa_tester"),
        OnTextMention(text="HANDOFF_TO_CODER", target="coder"),
    ],
    max_turns=200,
    max_tokens=60000,
    timeout_seconds=6000,
)

# ── Stage 3: Create PR ────────────────────────────────────────────

git_push_pr = Agent(
    name="git_push_pr",
    model=MODEL,
    credentials=["GITHUB_TOKEN", "GH_TOKEN"],
    instructions="""\
You are a PR creator. The branch is already pushed.
Run: gh pr create --repo <REPO> --base main --head <BRANCH> --title "..." --body "..."
Output the PR URL.
""",
    cli_commands=True,
    cli_allowed_commands=["gh"],
)

# ── Pipeline ──────────────────────────────────────────────────────

pipeline = git_fetch_issues >> coding_qa >> git_push_pr

if __name__ == "__main__":
    with AgentRuntime() as rt:
        # Deploy: push definition to server (idempotent — safe to call every startup).
        # Can also be done via CLI: agentspan deploy examples.production.github_coding_agent
        rt.deploy(pipeline)

        # Option A: Serve workers (production — blocks forever, run from outside)
        rt.serve(pipeline)

        # Option B: Run directly (quick test — uncomment below, comment out serve above)
        # rt.run() handles deploy + workers internally, no serve needed.
        # result = rt.run(pipeline, "Pick an open issue and create a PR.", timeout=240000)
        # result.print_result()
        #
        # Or trigger a deployed agent by name from any process:
        # result = rt.run("github_pipeline", "Pick an open issue and create a PR.")
        # result.print_result()

File: sdk/typescript/examples/61-github-coding-agent-chained.ts
/**
 * 61 - GitHub Coding Agent (Chained) — conditional sequential pipeline.
 *
 * Demonstrates:
 *   - Sequential pipeline with gate (conditional execution)
 *   - SWARM orchestration nested inside a pipeline stage
 *   - cliCommands for stages that only run CLI tools
 *   - localCodeExecution for stages that write/run code
 *
 * Architecture:
 *   pipeline = gitFetchIssues >> codingQA >> gitPushPR
 *
 * Requirements:
 *   - Conductor server running
 *   - AGENTSPAN_SERVER_URL=http://localhost:8080/api as environment variable
 *   - gh CLI authenticated
 */

import { Agent, AgentRuntime, OnTextMention, TextGate } from '../src/index.js';

const REPO = 'agentspan/codingexamples';
const MODEL = 'anthropic/claude-sonnet-4-6';

// -- Stage 1: Fetch issues ---------------------------------------------------

const gitFetchIssues = new Agent({
  name: 'git_fetch_issues',
  model: MODEL,
  instructions:
    `You are a GitHub issue fetcher.\n\n` +
    `1. List the 5 most recent open issues on ${REPO} (include number, title, body).\n` +
    `2. If there are NO open issues, output exactly: NO_OPEN_ISSUES\n` +
    `3. Otherwise pick the most suitable issue, then:\n` +
    `   - Create a temp dir: mktemp -d /tmp/fetch-XXXXXXXX\n` +
    `   - Clone ${REPO} into that dir and create branch fix/issue-<NUMBER>\n` +
    `   - Push the branch to origin immediately\n` +
    `   - Delete the temp dir.\n` +
    `   - Output ONLY these lines:\n` +
    `       REPO: ${REPO}\n` +
    `       BRANCH: fix/issue-<NUMBER>\n` +
    `       ISSUE: #<NUMBER> <title>\n` +
    `       SUMMARY: <one-sentence description>`,
  cliConfig: { enabled: true, allowedCommands: ['gh', 'git', 'mktemp', 'rm'] },
  maxTurns: 20,
  gate: new TextGate({ text: 'NO_OPEN_ISSUES' }),
});

// -- Stage 2: Coding + QA (SWARM) -------------------------------------------

const coderStage = new Agent({
  name: 'coder',
  model: MODEL,
  maxTokens: 60000,
  instructions:
    'You are a senior developer. Your task description contains REPO, BRANCH, ISSUE, and SUMMARY.\n\n' +
    '1. Create a fresh temp dir: mktemp -d /tmp/coder-XXXXXXXX\n' +
    '2. Clone the repo and check out the branch\n' +
    '3. Implement the fix described in ISSUE/SUMMARY\n' +
    '4. Commit your changes with a descriptive message\n' +
    '5. Push: git push origin <BRANCH>\n' +
    '6. Delete the temp dir\n' +
    '7. Say HANDOFF_TO_QA followed by REPO/BRANCH/CHANGES lines',
  codeExecutionConfig: { enabled: true },
});

const qaStage = new Agent({
  name: 'qa_tester',
  model: MODEL,
  instructions:
    'You are a QA engineer. Your task description contains REPO, BRANCH, and CHANGES.\n\n' +
    '1. Create a fresh temp dir and clone the repo/branch\n' +
    '2. Review the changed files and run tests\n' +
    '3. Delete the temp dir\n' +
    '4. If bugs: say HANDOFF_TO_CODER with details\n' +
    '5. If good: say QA_APPROVED followed by REPO/BRANCH/SUMMARY lines',
  codeExecutionConfig: { enabled: true },
  maxTokens: 60000,
  maxTurns: 5,
});

const codingQA = new Agent({
  name: 'coding_qa',
  model: MODEL,
  instructions:
    'Your task description contains REPO, BRANCH, ISSUE, and SUMMARY. ' +
    'Delegate to coder to implement the fix, passing REPO, BRANCH, and the task details. ' +
    'Once coder completes, delegate to qa_tester. ' +
    'If QA does not pass, send it back to coder to fix. ' +
    'When QA approves, output ONLY these lines:\n' +
    '  REPO: <repo>\n' +
    '  BRANCH: <branch>\n' +
    '  SUMMARY: <what was implemented and verified>',
  agents: [coderStage, qaStage],
  strategy: 'swarm',
  handoffs: [
    new OnTextMention({ text: 'HANDOFF_TO_QA', target: 'qa_tester' }),
    new OnTextMention({ text: 'HANDOFF_TO_CODER', target: 'coder' }),
  ],
  maxTurns: 200,
  maxTokens: 60000,
  timeoutSeconds: 6000,
});

// -- Stage 3: Create PR ------------------------------------------------------

const gitPushPR = new Agent({
  name: 'git_push_pr',
  model: MODEL,
  instructions:
    'You are a GitHub PR creator. Your task description contains REPO, BRANCH, and SUMMARY.\n' +
    'The branch is already pushed -- your only job is to open a pull request.\n\n' +
    '1. Create the PR: gh pr create --repo <REPO> --base main --head <BRANCH> --title "<title>" --body "<summary>"\n' +
    '2. Output the PR URL.',
  cliConfig: { enabled: true, allowedCommands: ['gh'] },
  maxTokens: 60000,
  maxTurns: 10,
});

// -- Pipeline ----------------------------------------------------------------

const pipeline = gitFetchIssues.pipe(codingQA).pipe(gitPushPR);

// Just deploy (since serve is blocking, per the Python source)
const runtime = new AgentRuntime();
try {
  const info = await runtime.deploy(pipeline);
  console.log(`Deployed: ${info.agentName} -> ${info.workflowName}`);
  console.log('Pipeline deployed. Use runtime.serve() to start workers.');
} finally {
  await runtime.shutdown();
}
